src/charm.py

```diff
@@ -181,28 +173,29 @@ def _configure_amf(self, event: EventBase) -> None:  # noqa C901
                 return
         if not self._database_is_available():
             self.unit.status = WaitingStatus("Waiting for the amf database to be available")
-            event.defer()
             return
         if not self._get_database_info():
             self.unit.status = WaitingStatus("Waiting for AMF database info to be available")
-            event.defer()
             return
         if not self._nrf_requires.nrf_url:
             self.unit.status = WaitingStatus("Waiting for NRF data to be available")
-            event.defer()
             return
         if not self._amf_container.exists(path=CONFIG_DIR_PATH):
             self.unit.status = WaitingStatus("Waiting for storage to be attached")
-            event.defer()
             return
         if not _get_pod_ip():
             self.unit.status = WaitingStatus("Waiting for pod IP address to be available")
-            event.defer()
             return
+        if not self._private_key_is_stored():
+            self._generate_private_key()
         if not self._certificate_is_stored():
-            self.unit.status = WaitingStatus("Waiting for certificates to be stored")
-            event.defer()
-            return
+            if not self._csr_is_stored():
+                self._request_new_certificate()
+                self.unit.status = WaitingStatus("Waiting for certificates to be stored")
+                return
+            else:
+                for provider_certificate in self._certificates.get_assigned_certificates():
+                    self._store_certificate(certificate=provider_certificate.certificate)
```

Comment:
I'm unconvinced about this step here? Does this charm require multiple certs?

Reply:
No, it does not.  There should be one and only one

Reply:
Loop is back as we need to compare the current CSR against the list of certs